README.md

```diff
@@ -190,9 +191,12 @@ Section 3.2.2.
 * 4.1.4. `a` SHOULD publish an encrypted `group/exclude` message on `Ga` with
 the following fields in the message `content`:
   * 4.1.4.A. `type` equals the string `group/exclude`
-  * 4.1.4.B. `excludes` is an array of group member IDs (their root metafeed
-  IDs) excluded from `G`.  In this case `c` is the only excluded member, but
-  Section 4.1. supports excluding multiple members at once.
+  * 4.1.4.B. `excludes` is an array of objects with the shape
+  `{id, groupFeedId, sequence}` where `id` is the root metafeed ID of `c`,
+  `groupFeedId` is `Gc`'s ID, and `sequence` is the sequence number of the last
+  message `a` possesses from `Gc`.  In this case `c` is the only excluded
+  member, so we only have one `{id, groupFeedId, sequence}` object but Section
+  4.1. also supports excluding multiple members at once.
```

Comment:
We should be able to (i think?) calculate this technically by looking at the group tangle but I guess this way is both clearer and more efficient.

Reply:
A really fun bug would be if the excluder starts creating this message, but then inbetween that and creating the tangle for the message, receives another message from the excluded person, pointing the group tangle at that message in the `tangles.group.previous`. Then the announce of this `sequence` and the group tangle would disagree on which messages to include.

I guess a way of preventing that would be for the excluder to stop replicating the excluded person before starting to create this message.

Reply:
Minus the tangle thing, I wouldn't see a problem. The admin saw "the last straw" from Oscar and decided that Oscar has to leave. It doesn't matter if Oscar keeps on posting because "the last straw" message is the one that the admin decided is the last one.

But since you mentioned tangles and their linking, and since I don't understand tangles (I thought I did), I'm like :shrug: and need someone to explain like I'm five. Maybe @mixmix ?

Reply:
yeah but if we later on add e.g. tangle sync, then people using that would end up with a different set of messages than people syncing whole feeds.

either way this is just in case an implementation has what's essentially a bug.

maybe just add a warning, something like

"implementers SHOULD (must?) be careful to not include newer messages than `sequence` in any tangle."

Reply:
@staltz did you get to understanding tangles? did my scuttelbutt write-up help?

Hmm... if the tangle thing is gonna be used/ useful we have an interesting edge case:
- Staltz excludes `{ id: Oscar, groupFeedId: OscarG,  sequence: 234 }`
- Oscar published message in `OscarG` with `sequence: 235`
- Mix replicates Oscar, and extends the group tangle off of that message of Oscars, then replicates Staltz and is like "oops I included a message after the exclusion in the group tangle").

We aren't using the tangle ordering (`tangles.group.previous`) thing yet for anything... so this is quite hard to design for. We mainly use `tangles.group.root` to find messages in a particular group....

Gut feeling is say "fuck it" and just continue on.... but if people receive Staltz's exclusion message first they will stop replicating and then have a gap in their tangle... hmmm.
Hard

Reply:
In the context of group feeds and epoch feeds, what problem *exactly* do tangles solve?

Reply:
I guess in the situation that mix describes, trusting the tangles instead of the announced `sequence` is more forgiving/appropriate for eventual consistency. It'd mean that it's ok to refer to an excluded feed after the moment it's been excluded, at least if the reason for referring to it is because you haven't heard of the exclusion yet.

Reply:
As I see it, based on what Mix said in Signal, tangles are solving an imaginary/future problem, but in this case with this PR they are *creating* a new problem. IMO we could just remove tangles here, why are we shooting ourselves in the foot? [KISS](https://en.wikipedia.org/wiki/KISS_principle) maybe

Reply:
No, in this situation

> Mix replicates Oscar, and extends the group tangle off of that message of Oscars, then replicates Staltz and is like "oops I included a message after the exclusion in the group tangle").

even without tangles, Mix might be replying to Oscar's latest comment. But Staltz would never see that comment since he doesn't want to go beyond the `sequence` he announce before. This would break discussion threads. Tangles in this case doesn't *create* the problem, it's a *method* of dealing with the problem that is eventual consistency.

Reply:
https://github.com/ssbc/ssb-group-exclusion-spec/pull/19

Reply:
> Mix might be replying to Oscar's latest comment. But Staltz would never see that comment

I could see Mix's comment, I just won't see what did Oscar say that Mix is replying to. I think that's an okay compromise to make.

By allowing Oscar to be "eventually excluded", we open up a few abuse channels too, because Oscar knows he's excluded but still has his final words to make, and at that point he might not give a shit about the code of conduct anymore. I think that's a worse outcome than discussion threads looking slightly punched with holes.

Reply:
We already don't make much of/any effort to deal with bad actors once they're inside a group. And not accomodating for eventual consistency would break a core design choice of scuttlebutt. It's not necessarily "slightly punched with holes", it might as well be long discussions. It'd also break i think one of our design principles for groups, that everyone should have the same set of messages etc. That's e.g. the reason we ignore blocks in groups.

Reply:
@mixmix @arj03 Help us out here? We're stuck